ngsPETSc/plex.py

```diff
@@ -16,11 +15,184 @@ class ngs:
         class comp:
             "dummy class"
             Mesh = type(None)
+try:
+    from numba import jit
+    numba = True 
+except:
+    jit = None
+    numba = False
 
 FACE_SETS_LABEL = "Face Sets"
 CELL_SETS_LABEL = "Cell Sets"
 EDGE_SETS_LABEL = "Edge Sets"
 
+if numba:
+    @jit(nopython=True, cache=True)
+    def build2DCells(coordinates, sIndicies, cell_indicies, start_end, vStart):
+        cStart = start_end[0]
+        cEnd = start_end[1]
+        cells = np.zeros((cell_indicies.shape[0], 3))
+        for i in range(cStart,cEnd):
+            sIndex = sIndicies[i]
+            if len(sIndex)==3:
+                cell = list(set(cell_indicies[i]))
+                A = np.zeros((2,2))
+                A[0,0] = (coordinates[cell[1]]-coordinates[cell[0]])[0]
+                A[0,1] = (coordinates[cell[1]]-coordinates[cell[0]])[1]
+                A[1,0] = (coordinates[cell[2]]-coordinates[cell[1]])[0]
+                A[1,1] = (coordinates[cell[2]]-coordinates[cell[1]])[1]
+                if np.linalg.det(A) > 0:
+                    cells[i] = cell
+                else:
+                    cells[i] = np.array([cell[0],cell[2],cell[1]])
+            else:
+                raise RuntimeError("We only support triangles.")
+        return cells
+
+    @jit(nopython=True, cache=True)
+    def build3DCells(coordinates, sIndicies, cell_indicies, start_end, vStart):
+        cStart = start_end[0]
+        cEnd = start_end[1]
+        cells = np.zeros((cell_indicies.shape[0], 4))
+        for i in range(cStart,cEnd):
+            sIndex = sIndicies[i]
+            if len(sIndex)==4:
+                cell = list(set(cell_indicies[i]))
```

Comment:
What is the role of casting into a `set`? Are we ever going to get repeated vertices or do we only want to sort the vertices?

Reply:
Remove duplicates